File: day12/day12a.py
#!/usr/bin/python3

import math

with open('day12.txt', 'r') as f:
    lines = f.readlines()
    posX = 0
    posY = 0
    theta = 0
    for line in lines:
        action = line[0]
        value = int(line[1:])
        if action == 'N':
            posY += value
        elif action == 'S':
            posY -= value
        elif action == 'E':
            posX += value
        elif action == 'W':
            posX -= value
        elif action == 'L':
            theta += value
        elif action == 'R':
            theta -= value
        elif action == 'F':
            posX += value * math.cos(math.radians(theta))
            posY += value * math.sin(math.radians(theta))

    print(round(abs(posX) + abs(posY)))

File: day12/day12a.cpp
#include <cmath>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

int main() {
  // Read input from file and sort
  std::ifstream file("day12.txt");
  std::string line;
  char action;
  int value;

  double posX = 0;
  double posY = 0;
  double theta = 0;

  while(std::getline(file, line)) {
    std::stringstream line_ss(line);
    line_ss >> action;
    line_ss >> value;

    switch(action) {
      case 'N':
        posY += value;
        break;
      case 'S':
        posY -= value;
        break;
      case 'E':
        posX += value;
        break;
      case 'W':
        posX -= value;
        break;
      case 'L':
        theta += value;
        break;
      case 'R':
        theta -= value;
        break;
      case 'F':
        posX += value * std::cos(theta * 3.1415926535 / 180.0);
        posY += value * std::sin(theta * 3.1415926535 / 180.0);
        break;
    }
  }

  std::cout << static_cast<int>(std::round(std::abs(posX))) +
               static_cast<int>(std::round(std::abs(posY))) << std::endl;
  return 0;
}
